File: src/s0076_minimum_window_substring.rs
/**
 * [76] Minimum Window Substring
 *
 * Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).
 *
 * Example:
 *
 *
 * Input: S = "ADOBECODEBANC", T = "ABC"
 * Output: "BANC"
 *
 *
 * Note:
 *
 *
 * 	If there is no such window in S that covers all characters in T, return the empty string "".
 * 	If there is such window, you are guaranteed that there will always be only one unique minimum window in S.
 *
 *
 */

// problem: https://leetcode.cn/problems/minimum-window-substring/

// submission codes start here
use std::collections::HashMap;

pub struct Solution;

impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        if s.len() == 1 && t.len() == 1 && s == t { return s; }

        // 记录最小覆盖子串的起始索引及长度
        // let mut start = 0;
        // let mut len = usize::MAX;
        let mut ret = String::from("");

        let (mut left, mut right, mut valid) = (0usize, 0usize, 0);
        let mut needs = HashMap::new();
        let mut first = true;
        t.chars().for_each(|c| { needs.entry(c).and_modify(|counter| *counter += 1).or_insert(1); } );

        let mut window = HashMap::new();

        while right < s.len() {
            let r = s.chars().nth(right).unwrap();
            right += 1;

            if needs.contains_key(&r) {
                window.entry(r).and_modify(|counter| *counter +=1 ).or_insert(1);
                if window.get(&r) == needs.get(&r) { valid += 1; }
            }

            while valid == needs.len() {
                // if right - left < len {
                //     start = left;
                //     len = right - left;
                // }
                if first { 
                    ret = s[left..right].to_string(); 
                    first = false;
                } else {
                    if (&s[left..right]).len() < ret.len() { ret = s[left..right].to_string(); }
                }

                let l = s.chars().nth(left).unwrap();
                left += 1;

                if needs.contains_key(&l) {
                    if window.get(&l) == needs.get(&l) { valid -= 1; }
                    window.entry(l).and_modify(|counter| *counter -=1 );
                }
            }
            
        }

        ret
        // return if len == usize::MAX { String::from("") } else { s[start..(start + len)].to_string() }
    }
}

#[test]
fn test_76() {
    assert_eq!(
        Solution::min_window(String::from("a"), String::from("a")),
        "a".to_string()
    );

    let s = String::from("ADOBECODEBANC");
    let t = String::from("ABC");
    assert_eq!(
        Solution::min_window(s, t),
        "BANC".to_string()
    );
}

File: cpp/sliding-window/s0076_minimum_window_substring.cpp
// problem: https://leetcode.cn/problems/minimum-window-substring/
// 示例 1:
// 输入: s = "ADOBECODEBANC", t = "ABC"
// 输出: "BANC"
// 解释: 最小覆盖子串 "BANC" 包含来自字符串 t 的 'A'、'B' 和 'C'.
// 
// 示例 2:
// 输入: s = "a", t = "a"
// 输出: "a"
// 解释: 整个字符串 s 是最小覆盖子串.
// 
// 示例 3:
// 输入: s = "a", t = "aa"
// 输出: ""
// 解释: t 中两个字符 'a' 均应包含在 s 的子串中.
// 因此没有符合条件的子字符串，返回空字符串.
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        std::transform(s.begin(), s.end(), s.begin(), [](unsigned char ch) { return std::tolower(ch); });
        std::transform(t.begin(), t.end(), t.begin(), [](unsigned char ch) { return std::tolower(ch); });
        unordered_map<char, int> needs, umap;
        for (auto ch : t) needs[ch]++;
        int left = 0, right = 0;
        int valid = 0;
        int start = 0, min_len = INT_MAX;

        while (right < s.size()) {
            char rc = s[right];
            right ++;
            if (needs.count(rc)) {
                umap[rc] ++;
                if (umap[rc] == needs[rc]) valid ++;
            }
            while (valid == needs.size()) {
                if (min_len > right - left) {
                    start = left;
                    min_len = right - left;
                }
                char lc = s[left];
                left ++;
                if (needs.count(lc)) {
                    if (needs[lc] == umap[lc]) valid --;
                    umap[lc] --;
                }
            }
        }

        return min_len == INT_MAX ? "" : s.substr(start, min_len);
    }
};

int main()
{
    return 0;
}